flowd.8.in
.\" $Id$
.\"
.\" Copyright (c) 2004 Damien Miller <djm@mindrot.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd July 30, 2004
.Dt FLOWD 8
.Os
.Sh NAME
.Nm flowd
.Nd NetFlow collector daemon
.Sh SYNOPSIS
.Nm flowd
.Op Fl hd
.Op Fl f Ar config_file
.Ar command
.Sh DESCRIPTION
.Nm
is a small NetFlow collector daemon capable of understanding Cisco NetFlow 
version 1 and version 5 packet formats.
.Nm
supports filtering and tagging of received flows before they are stored on 
disk, using a filter syntax similar to the OpenBSD PF packet filter.
The on-disk format is flexible in that it allows selection of which packet
fields are recorded, so logs may be made very compact.
.Pp
By default,
.Nm
will obtain its configuration from the 
.Xr flowd.conf 5
file, which typically resides in
.Pa @CONFPATH@/flowd.conf .
However, a different configuration file may be specified on the commandline
using the 
.Fl d
option.
.Pp
.Nm
may be terminated by sending it a 
.Dv SIGTERM
signal.
Upon receipt of a
.Dv SIGUSR1
.Nm
will close and reopen its logfile.
This is useful for rotating logs.
When
.Nm
receives a 
.Dv SIGHUP
it will re-read its configuration and re-open its logfile.
Some basic runtime statistics will be logged when 
.Nm
is signalled with
.Dv SIGUSR2
or
.Dv SIGINFO .
.Pp
The command-line options are as follows:
.Bl -tag -width Ds
.It Fl d
Places
.Nm
in debugging mode.
In this mode,
.Nm
will not fork to the background and will print verbose diagnostics, including
details of each flow received.
.It Fl f Ar config_file
Specify an alternate location for the configuration file
Default is 
.Pa @CONFPATH@/flowd.conf
.It Fl h
Displays commandline usage information.
.El
.Sh AUTHORS
Damien Miller <djm@mindrot.org>
.Sh SEE ALSO
.Xr flowd-reader 8
.Xr flowd.conf 5


ChangeLog
20041030
 - (djm) Rework flow reading C API to be more friendly to wrapping into Perl
   and Python: add functions to convert a serialised flow record into a
   struct store_flow_complete.
 - (djm) Skip CRC32 calculations when aren't reading/writing a CRC32, saves a
   little time.
 - (djm) Add a serialisation function to match the new deserialisation function

20041011
 - (djm) Kill whitespace at EOL, improve nf9 debugging
 - (djm) Add option to keep flowd in foreground without verbose debugging; 
   patch from lars AT unet.net.ph
 - (djm) Document new -g and existing -D options

20040929
 - (djm) Don't refuse to write flows lacking gateway addr

20040927
 - (djm) In flow formattinhg function (C, Perl and Python), surround all
   addresses with square brackets so ports may be seen in IPv6 addresses.
   Spotted by Chris Gascoigne

20040924
 - (djm) Tidy README and mention v.7 and v.9 support
 - (djm) Recommend use of "flow source" in flowd.conf for NetFlow 9
 - (djm) Add tree code for recent changes
 - (djm) Store agent address for v.9 flows
 - (djm) Add some tools to the tools/ directory
 - (djm) Add tools/ directory to RPM
 - (djm) Make RPM tools subpackage
 - (djm) Release flowd-0.6

20040921
 - (djm) Implement NetFlow v.9 support. Some more work to do yet, but it is 
   basically functional
 - (djm) Bugfix nf v.9: we never updated the peer record
 - (djm) Tidy nv v.9 code: factor out data flowset to store flow conversion and
   don't process any flows unless the entire flowset is valid.

20040920
 - (djm) Add new "flow source" directive, to specify hosts that we are willing
   to receive flows from
 - (djm) Split peer tracking code out into separate file

20040916
 - (djm) Fix transposed flow limits for v.1 vs v.5 flows; found by 
   jon AT exalia.com
 - (djm) Generate version number from autoconf, rather than hardcoding old (!)
   value in Makefile; found by jon AT exalia.com
 - (djm) Fix permissions on generated setup.py
 - (djm) Release flowd-0.5
 - (djm) Add packet dumping code (disabled by default)
 - (djm) Add untested NetFlow v.7 parsing code
 - (djm) Add peer tracking and stateholding code, will be needed to support 
   NetFlow v.9 and IPFIX

20040914
 - (djm) Use a table-based parser in Flowd.pm, like the python code

20040906
 - (djm) Actually check for strlcpy and strlcat in configure

20040908
 - (djm) Fix parsing bug that could suppress errors for invalid addresses in
   filter rules

20040906
 - (djm) Reformat README, upgrade from beta to production quality and mention
   PLATFORMS file
 - (djm) Mention FreeBSD and Redhat 7.3 in PLATFORMS file
 - (djm) Skip CRC calculation in Perl and Python code when there is no CRC 
   in the record. Speeds up processing of CRC-less files considerably
 - (djm) Generate Flowd.pm, flowd.py and setup.py from template files, saving
   me the necessity of editing each whenever the version changes

20040823
 - (djm) NetFlow v.5 allows 30 flows per packet, not 24
 - (djm) Call tzset() so unpriv child gets log timestamps right
 - (djm) Missing htons() in error path

20040819
 - (djm) Fixes to Flowd.pm for Redhat 7.3 (Perl 5.6.1):
   - "use constant" syntax
   - Pull PF_INET6 from Socket6, not Socket
   - Compatible Math::BigInt code
   - Remove leading '+' from Math::BigInts when printing
   - Avoid Math::BigInt altogether for counters < 2^32
 - (djm) Don't refuse to load configs without filter rules
 - (djm) Support "store SRC_ADDR" and "store DST_ADDR" aliases
 - (djm) Fix flowd.conf store example
 - (djm) Crank versions to 0.4.1

20040817
 - (djm) Make sure we reinstate signal handers after invocation, fixes 
         exit on 2nd reconfigure on Solaris
 - (djm) Don't rewrite pidfile if its path hasn't changed - avoids race 
         condition false-failures in "while [ 1 ] ; do kill ... ; done" 
         stress test
 - (djm) Improve PLATFORMS file a little
 - (djm) Release flowd 0.4
 - (djm) Unbreak spec file for Redhat 9

20040816
 - (djm) Pull in some portability fixes from portable OpenSSH
 - (djm) Crank version
 - (djm) Portability fixes for Redhat 9
 - (djm) Make Python API optional in RPM spec as it is broken on Redhat 9
 - (djm) Add some standard logging code from softflowd, that does stderr right
         and is a little more portable
 - (djm) Add PLATFORMS notes file
 - (djm) Lots of portability goop to get flowd compiling of Solaris 9
 - (djm) Reflect Solaris support in PLATFORMS and INSTALL
 
20040813
 - (djm) Bring python inteface up to the same level as the perl and C ones
 - (djm) Print tcp_flags as two hex digits, in C API
 - (djm) Improve the README, mention Python API
 - (djm) Reorganise signal handers, add SIGUSR1 for reopen logfiles
 - (djm) Document signals and tidy
 - (djm) Portability and warning fixes
 - (djm) Add python install script and mention it in INSTALL
 - (djm) Enable reload method in initscript, now that it works
 - (djm) Rework RPM spec to have perl and python modules in subpackages
 - (djm) Release flowd 0.3

20040812
 - (djm) Parse and verify config in a subprocess, avoiding all memory leaks
   and side effects.
 - (djm) Activate runtime reconfigure
 - (djm) Store address families for each address in flow
 - (djm) Add start of a python interface to the flow log format

20040810
 - (djm) Add finish() method to perl API, to close flow log
 - (djm) Fix filter format_rule for rules that match ports but not addrs
 - (djm) Set IPV6_V6ONLY on AF_INET6 sockets - avoids stupid mapped address crap
 - (djm) Don't leak fd on logfile reopen
 - (djm) Move listener open function to privsep.c, in preparation for runtime 
   reconfigure support
 - (djm) Implement runtime reconfigure, currently disabled because of memory 
   leaks in config parsing code

20040804
 - (djm) Don't clobber existing configuration file
 - (djm) Rework the storage format a little, so we can be more compact when
   storing a few fields - NB this is an incompatible format change

flowd.py
# Copyright (c) 2004 Damien Miller <djm@mindrot.org>
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

# $Id$

VERSION = "0.8"

import struct
import time
import socket
import sys
import flowd_serialiser

def iso_time(secs, utc = 0):
	if utc:
		tm = time.gmtime(secs)
	else:
		tm = time.localtime(secs)

	return "%04u-%02u-%02uT%02u:%02u:%02u" % tm[:6]

def interval_time(t):
	intervals = [	[ "s", 60 ], [ "m", 60 ], [ "h", 24 ],
			[ "d", 7 ], [ "w", 52 ]]
	ret = ""
	for interval in intervals:
		r = t % interval[1]
		t = int(t / interval[1])
		if r != 0 or interval[0] == "s":
			ret = "%u%s%s" % (r, interval[0], ret)
	if t > 0:
		ret = "%uy%s", (t, ret)

	return ret


def interval_time_ms(tms):
	return "%s.%03u" % ( interval_time(int(tms / 1000)), tms % 1000 )

class log:
	def __init__(self, path, mode = "r", start_time = None):
		self.path = path
		self.mode = mode
		if mode == "r":
			self.flow_file = open(path, "rb")
			# Read header
			hdr = self.flow_file.read(16)
			if len(hdr) != 16:
				raise ValueError, "Short read on flow header"
			(self.magic, self.version, self.start_time, \
			 self.flags) = struct.unpack(">IIII", hdr)

			if self.magic != 0x012cf047:
				raise ValueError, "Bad magic"
			if self.version != 0x00000002:
				raise ValueError, "Unsupported version"
		elif mode == "w" or mode == "a":
			self.flow_file = open(path, mode + "b")
			self.flow_file.seek(0, 2)
			if self.flow_file.tell() > 0:
				return

			# Write header
			self.magic = 0x012cf047
			self.version = 0x02
			self.flags = 0x00
			self.start_time = start_time
			if start_time is None:
				self.start_time = int(time.time())
			hdr = struct.pack(">IIII", self.magic, self.version, \
			    self.start_time, self.flags)
			if len(hdr) != 16:
				raise ValueError, "Internal error: bad header"
			self.flow_file.write(hdr)
			self.flow_file.flush()
		else:
			raise ValueError, "Invalid mode value";

	def finish(self):
		self.flow_file.close()
		self.flow_file = None

	def readflow(self):
		try:
			f = flow()
			f.from_file(self.flow_file)
		except EOFError:
			f = None
		return f

	def writeflow(self, flow):
		flow.to_file(self.flow_file)

class flow:
	TAG			= 0x00000001
	RECV_TIME		= 0x00000002
	PROTO_FLAGS_TOS		= 0x00000004
	AGENT_ADDR4		= 0x00000008
	AGENT_ADDR6		= 0x00000010
	SRC_ADDR4		= 0x00000020
	SRC_ADDR6		= 0x00000040
	DST_ADDR4		= 0x00000080
	DST_ADDR6		= 0x00000100
	GATEWAY_ADDR4		= 0x00000200
	GATEWAY_ADDR6		= 0x00000400
	SRCDST_PORT		= 0x00000800
	PACKETS			= 0x00001000
	OCTETS			= 0x00002000
	IF_INDICES		= 0x00004000
	AGENT_INFO		= 0x00008000
	FLOW_TIMES		= 0x00010000
	AS_INFO			= 0x00020000
	FLOW_ENGINE_INFO	= 0x00040000
	CRC32			= 0x40000000

	# Some useful combinations
	AGENT_ADDR		= 0x00000018
	SRC_ADDR		= 0x00000060
	DST_ADDR		= 0x00000180
	SRCDST_ADDR		= 0x000001e0
	GATEWAY_ADDR		= 0x00000600
	BRIEF			= 0x000039ff
	ALL			= 0x4007ffff

	def __init__(self):
		self.fields = { "fields" : 0 }

	def from_file(self, flow_file):
		# Read flow header
		needlen = flowd_serialiser.header_len()
		hdr = flow_file.read(needlen)
		if len(hdr) == 0:
			raise EOFError
		if len(hdr) != needlen:
			raise ValueError, "Short read on flow header"

		needlen = flowd_serialiser.flow_len(hdr)
		flow = flow_file.read(needlen)
		if len(flow) == 0:
			raise EOFError
		if len(flow) != needlen:
			raise ValueError, "Short read on flow data"

		self.fields = flowd_serialiser.deserialise(hdr + flow)

	def to_file(self, flow_file, field_mask = 0xffffffffL):
		flow = flowd_serialiser.serialise(self.fields, field_mask)
		flow_file.write(flow)
		flow_file.flush()

	def format(self, field_mask = BRIEF, utc = 0):
		fields = self.fields["fields"] & field_mask
		ret = "FLOW "

		if fields & self.__class__.TAG != 0:
			ret = ret + "tag %u " % self.fields["tag"]
		if fields & self.__class__.RECV_TIME != 0:
			ret = ret + "recv_time %s " % \
			    iso_time(self.fields["recv_secs"], utc)
		if fields & self.__class__.PROTO_FLAGS_TOS != 0:
			ret = ret + "proto %u " % self.fields["protocol"]
			ret = ret + "tcpflags %02x " % self.fields["tcp_flags"]
			ret = ret + "tos %02x " % self.fields["tos"]
		if fields & self.__class__.AGENT_ADDR != 0:
			ret = ret + "agent [%s] " % self.fields["agent_addr"]
		if fields & self.__class__.SRC_ADDR != 0:
			ret = ret + "src [%s]" % self.fields["src_addr"];
			if fields & self.__class__.SRCDST_PORT != 0:
				ret = ret + ":%u" % self.fields["src_port"];
			ret = ret + " ";
		if fields & self.__class__.DST_ADDR != 0:
			ret = ret + "dst [%s]" % self.fields["dst_addr"];
			if fields & self.__class__.SRCDST_PORT != 0:
				ret = ret + ":%u" % self.fields["dst_port"];
			ret = ret + " ";
		if fields & self.__class__.GATEWAY_ADDR != 0:
			ret = ret + "gateway [%s] " % \
			    self.fields["gateway_addr"];
		if fields & self.__class__.PACKETS != 0:
			ret = ret + "packets %s " % self.fields["flow_packets"];
		if fields & self.__class__.OCTETS != 0:
			ret = ret + "octets %s " % self.fields["flow_octets"];
		if fields & self.__class__.IF_INDICES != 0:
			ret = ret + "in_if %u " % self.fields["if_index_in"];
			ret = ret + "out_if %u " % self.fields["if_index_out"];
		if fields & self.__class__.AGENT_INFO != 0:
			ret = ret + "sys_uptime_ms %s " % \
			    interval_time_ms(self.fields["sys_uptime_ms"]);
			ret = ret + "time_sec %s " % \
			    iso_time(self.fields["time_sec"], utc);
			ret = ret + "time_nanosec %u " % \
			    self.fields["time_nanosec"];
			ret = ret + "netflow ver %u " % \
				self.fields["netflow_version"];
		if fields & self.__class__.FLOW_TIMES != 0:
			ret = ret + "flow_start %s " % \
			    interval_time_ms(self.fields["flow_start"]);
			ret = ret + "flow_finish %s " % \
			    interval_time_ms(self.fields["flow_finish"]);
		if fields & self.__class__.AS_INFO != 0:
			ret = ret + "src_AS %u " % self.fields["src_as"];
			ret = ret + "src_masklen %u " % \
			    self.fields["src_masklen"];
			ret = ret + "dst_AS %u " % self.fields["dst_as"];
			ret = ret + "dst_masklen %u " % \
			    self.fields["dst_masklen"];
		if fields & self.__class__.FLOW_ENGINE_INFO != 0:
			ret = ret + "engine_type %u " % \
			    self.fields["engine_type"];
			ret = ret + "engine_id %u " % self.fields["engine_id"];
			ret = ret + "seq %u " % self.fields["flow_sequence"];
		if fields & self.__class__.CRC32 != 0:
			ret = ret + "crc32 %08x " % self.fields["crc"];

		return ret;
